In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

pd.pandas.set_option('display.max_columns', None)

# ML Modeling using Mutual Information dataset

In [ ]:
dataset_MI = pd.read_csv("C:\Technocolabs Internship\Prosper Project\Financial-Risk-Analysis-and-Prediction-of-online-P2P-Lending\EDA_MI_dataset.csv")

In [4]:
dataset_MI.head()

Term  ListingCategory (numeric)  IsBorrowerHomeowner  CurrentlyInGroup  \
0 -0.404987                  -1.700178            -1.008984         -0.354485   
1 -0.404987                  -1.700178            -1.008984         -0.354485   
2 -0.404987                  -1.700178            -1.008984         -0.354485   
3 -0.404987                  -1.700178            -1.008984         -0.354485   
4 -0.404987                  -1.700178            -1.008984         -0.354485   

   CurrentCreditLines  InquiriesLast6Months  TotalInquiries  \
0            0.079384              0.337495        0.045835   
1            0.079384              0.337495        0.045835   
2            0.079384              0.337495        0.045835   
3            0.079384              0.337495        0.045835   
4            0.079384              0.337495        0.045835   

   CurrentDelinquencies  DelinquenciesLast7Years  PublicRecordsLast10Years  \
0             -0.509351                -0.673914                 -0.563129   
1             -0.509351                -0.673914                 -0.563129   
2             -0.509351                -0.673914                 -0.563129   
3             -0.509351                -0.673914                 -0.563129   
4             -0.509351                -0.673914                 -0.563129   

   TradesNeverDelinquent (percentage)  TradesOpenedLast6Months  \
0                            0.159427                -0.895144   
1                            0.159427                -0.895144   
2                            0.159427                -0.895144   
3                            0.159427                -0.895144   
4                            0.159427                -0.895144   

   IncomeVerifiable  LoanCurrentDaysDelinquent  LP_CollectionFees  \
0           0.28697                  -0.447962           0.277852   
1           0.28697                  -0.447962           0.277852   
2           0.28697                  -0.447962           0.277852   
3           0.28697                  -0.447962           0.277852   
4           0.28697                  -0.447962           0.277852   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
0              -0.300062            -0.259434   
1              -0.300062            -0.259434   
2              -0.300062            -0.259434   
3              -0.300062            -0.259434   
4              -0.300062            -0.259434   

   LP_NonPrincipalRecoverypayments  InvestmentFromFriendsCount  \
0                        -0.171621                   -0.138057   
1                        -0.171621                   -0.138057   
2                        -0.171621                   -0.138057   
3                        -0.171621                   -0.138057   
4                        -0.171621                   -0.138057   

   LoanOriginationYear  LoanOriginationMonth  ClosedDateYear  ClosedDateMonth  \
0            -2.410658              1.235732       -3.622081          1.22201   
1            -2.410658              1.235732       -3.622081          1.22201   
2            -2.410658              1.235732       -3.091549         -1.50196   
3            -2.410658              1.235732       -3.091549         -1.50196   
4            -2.410658              1.235732       -3.622081          1.22201   

   ListingCreationYear  DateCreditPulledYear  DateCreditPulledMonth  \
0            -2.400822             -2.398311               1.245448   
1            -2.400822             -2.398311               1.245448   
2            -2.400822             -2.398311               1.245448   
3            -2.400822             -2.398311               1.245448   
4            -2.400822             -2.398311               1.245448   

   LoanOriginationQuarterYear  LoanOriginationQuarterMonth  \
0                   -2.410658                     1.274029   
1                   -2.410658                     1.274029   
2                   -2.410658                     1.274029   
3     

In [5]:
dataset_MI.shape

(113937, 41)

In [6]:
dataset_MI['IsDefault'].value_counts()

IsDefault
0    96927
1    17010
Name: count, dtype: int64

In [7]:
print(dataset_MI['IsDefault'].value_counts(normalize=True))

IsDefault
0    0.850707
1    0.149293
Name: proportion, dtype: float64


The dataset is highly imbalanced. So, performing Oversampling before applying the ML models.

In [8]:
# Independent and dependent variables

X_MI = dataset_MI.drop(['IsDefault'], axis=1)
y_MI = dataset_MI['IsDefault']

In [9]:
X_MI.head()

Term  ListingCategory (numeric)  IsBorrowerHomeowner  CurrentlyInGroup  \
0 -0.404987                  -1.700178            -1.008984         -0.354485   
1 -0.404987                  -1.700178            -1.008984         -0.354485   
2 -0.404987                  -1.700178            -1.008984         -0.354485   
3 -0.404987                  -1.700178            -1.008984         -0.354485   
4 -0.404987                  -1.700178            -1.008984         -0.354485   

   CurrentCreditLines  InquiriesLast6Months  TotalInquiries  \
0            0.079384              0.337495        0.045835   
1            0.079384              0.337495        0.045835   
2            0.079384              0.337495        0.045835   
3            0.079384              0.337495        0.045835   
4            0.079384              0.337495        0.045835   

   CurrentDelinquencies  DelinquenciesLast7Years  PublicRecordsLast10Years  \
0             -0.509351                -0.673914                 -0.563129   
1             -0.509351                -0.673914                 -0.563129   
2             -0.509351                -0.673914                 -0.563129   
3             -0.509351                -0.673914                 -0.563129   
4             -0.509351                -0.673914                 -0.563129   

   TradesNeverDelinquent (percentage)  TradesOpenedLast6Months  \
0                            0.159427                -0.895144   
1                            0.159427                -0.895144   
2                            0.159427                -0.895144   
3                            0.159427                -0.895144   
4                            0.159427                -0.895144   

   IncomeVerifiable  LoanCurrentDaysDelinquent  LP_CollectionFees  \
0           0.28697                  -0.447962           0.277852   
1           0.28697                  -0.447962           0.277852   
2           0.28697                  -0.447962           0.277852   
3           0.28697                  -0.447962           0.277852   
4           0.28697                  -0.447962           0.277852   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
0              -0.300062            -0.259434   
1              -0.300062            -0.259434   
2              -0.300062            -0.259434   
3              -0.300062            -0.259434   
4              -0.300062            -0.259434   

   LP_NonPrincipalRecoverypayments  InvestmentFromFriendsCount  \
0                        -0.171621                   -0.138057   
1                        -0.171621                   -0.138057   
2                        -0.171621                   -0.138057   
3                        -0.171621                   -0.138057   
4                        -0.171621                   -0.138057   

   LoanOriginationYear  LoanOriginationMonth  ClosedDateYear  ClosedDateMonth  \
0            -2.410658              1.235732       -3.622081          1.22201   
1            -2.410658              1.235732       -3.622081          1.22201   
2            -2.410658              1.235732       -3.091549         -1.50196   
3            -2.410658              1.235732       -3.091549         -1.50196   
4            -2.410658              1.235732       -3.622081          1.22201   

   ListingCreationYear  DateCreditPulledYear  DateCreditPulledMonth  \
0            -2.400822             -2.398311               1.245448   
1            -2.400822             -2.398311               1.245448   
2            -2.400822             -2.398311               1.245448   
3            -2.400822             -2.398311               1.245448   
4            -2.400822             -2.398311               1.245448   

   LoanOriginationQuarterYear  LoanOriginationQuarterMonth  \
0                   -2.410658                     1.274029   
1                   -2.410658                     1.274029   
2                   -2.410658                     1.274029   
3     

In [10]:
y_MI.head()

0    0
1    0
2    0
3    0
4    0
Name: IsDefault, dtype: int64

In [11]:
# Splitting the dataset into train and test datasets

from sklearn.model_selection import train_test_split

X_train_MI, X_test_MI, y_train_MI, y_test_MI = train_test_split(X_MI, y_MI, test_size=0.2, random_state=42)

# OVERSAMPLING

In [12]:
from imblearn.combine import SMOTETomek, SMOTEENN
from imblearn.over_sampling import SMOTE
from imblearn.over_sampling import RandomOverSampler
from imblearn.over_sampling import ADASYN
from imblearn.over_sampling import BorderlineSMOTE
from collections import Counter

In [13]:
# os = SMOTE()
# X_train_upsampled, y_train_upsampled = os.fit_resample(X_train, y_train)

# print("The number of classes before oversampling: {}".format(Counter(y_train)))
# print("The number of classes after oversampling: {}".format(Counter(y_train_upsampled)))

In [14]:
# ros = RandomOverSampler(random_state=42)
# X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

# print("The number of classes before oversampling: {}".format(Counter(y_train)))
# print("The number of classes after oversampling: {}".format(Counter(y_resampled)))

In [15]:
os = ADASYN(random_state=42)
X_train_MI_upsampled, y_train_MI_upsampled = os.fit_resample(X_train_MI, y_train_MI)

print("The number of classes before oversampling: {}".format(Counter(y_train_MI)))
print("The number of classes after oversampling: {}".format(Counter(y_train_MI_upsampled)))

The number of classes before oversampling: Counter({0: 77569, 1: 13580})
The number of classes after oversampling: Counter({0: 77569, 1: 77438})


In [16]:
# os = BorderlineSMOTE(random_state=42)
# X_train_upsampled, y_train_upsampled = os.fit_resample(X_train, y_train)

# print("The number of classes before oversampling: {}".format(Counter(y_train)))
# print("The number of classes after oversampling: {}".format(Counter(y_train_upsampled)))

In [17]:
# os = SMOTEENN(random_state=42)
# X_train_upsampled, y_train_upsampled = os.fit_resample(X_train, y_train)

# print("The number of classes before oversampling: {}".format(Counter(y_train)))
# print("The number of classes after oversampling: {}".format(Counter(y_train_upsampled)))

In [18]:
print("Resampled class distribution:", np.bincount(y_train_MI_upsampled))

Resampled class distribution: [77569 77438]


# Applying Logistic Regression

In [19]:
from sklearn.linear_model import LogisticRegression

log_reg_model_MI = LogisticRegression()

log_reg_model_MI.fit(X_train_MI_upsampled, y_train_MI_upsampled)

LogisticRegression()

In [20]:
y_pred_log_reg_MI = log_reg_model_MI.predict(X_test_MI)

print(y_pred_log_reg_MI)

[1 0 1 ... 1 1 1]


In [1]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.metrics import roc_auc_score

In [22]:
# Evaluate the model
accuracy = accuracy_score(y_test_MI, y_pred_log_reg_MI)
print(f"Accuracy: {accuracy:.4f}")

# Print classification report
print("\nClassification Report:")
print(classification_report(y_test_MI, y_pred_log_reg_MI))

# Print confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test_MI, y_pred_log_reg_MI))

# Print confusion matrix
print("AUC score:")
print(roc_auc_score(y_test_MI, y_pred_log_reg_MI))

Accuracy: 0.4798

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.47      0.60     19358
           1       0.15      0.54      0.24      3430

    accuracy                           0.48     22788
   macro avg       0.50      0.51      0.42     22788
weighted avg       0.75      0.48      0.55     22788

Confusion Matrix:
[[ 9064 10294]
 [ 1561  1869]]
AUC score:
0.5065640741263961


# Applying Naive Bayes

In [23]:
from sklearn.naive_bayes import GaussianNB

In [24]:
gnb_model_MI = GaussianNB()

In [25]:
gnb_model_MI.fit(X_train_MI_upsampled, y_train_MI_upsampled)

GaussianNB()

In [26]:
y_pred_gnb_MI = gnb_model_MI.predict(X_test_MI)

In [27]:
# Evaluate the model
accuracy = accuracy_score(y_test_MI, y_pred_gnb_MI)
print(f"Accuracy: {accuracy:.4f}")

# Print classification report
print("\nClassification Report:")
print(classification_report(y_test_MI, y_pred_gnb_MI))

# Print confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test_MI, y_pred_gnb_MI))

# Print confusion matrix
print("AUC score:")
print(roc_auc_score(y_test_MI, y_pred_gnb_MI))

Accuracy: 0.2573

Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.15      0.26     19358
           1       0.15      0.84      0.25      3430

    accuracy                           0.26     22788
   macro avg       0.50      0.50      0.26     22788
weighted avg       0.74      0.26      0.26     22788

Confusion Matrix:
[[ 2997 16361]
 [  564  2866]]
AUC score:
0.4951941129498897


# Applying LightGBM

In [28]:
# from lightgbm import LGBMClassifier

import lightgbm as lgb

In [29]:
# Set the parameters for LightGBM

#Try dart for better accuracy

# lgb_model = LGBMClassifier(learning_rate=0.05, 
#             boosting_type = 'gbdt',    
#             objective ='binary',
#             num_leaves = 100,
#             max_depth = 30,
#             n_estimators=500,
#             class_weight='balanced')

lgb_model_MI = lgb.LGBMClassifier()

In [30]:
# Removing special characters from 'X_train_upsampled' columns

X_train_MI_upsampled.columns = X_train_MI_upsampled.columns.str.replace(r'[^\w]', '_', regex=True)

In [31]:
lgb_model_MI.fit(X_train_MI_upsampled, y_train_MI_upsampled)

[LightGBM] [Info] Number of positive: 77438, number of negative: 77569
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.028098 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 8877
[LightGBM] [Info] Number of data points in the train set: 155007, number of used features: 40
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.499577 -> initscore=-0.001690
[LightGBM] [Info] Start training from score -0.001690


LGBMClassifier()

In [32]:
# Prediction on test data
y_pred_lgbm_MI = lgb_model_MI.predict(X_test_MI)
# print(y_pred)

In [33]:
# Evaluate the model
accuracy = accuracy_score(y_test_MI, y_pred_lgbm_MI)
print(f"Accuracy: {accuracy:.4f}")

# Print classification report
print("\nClassification Report:")
print(classification_report(y_test_MI, y_pred_lgbm_MI))

# Print confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test_MI, y_pred_lgbm_MI))

# Print confusion matrix
print("AUC score:")
print(roc_auc_score(y_test_MI, y_pred_lgbm_MI))

Accuracy: 0.8495

Classification Report:
              precision    recall  f1-score   support

           0       0.85      1.00      0.92     19358
           1       0.00      0.00      0.00      3430

    accuracy                           0.85     22788
   macro avg       0.42      0.50      0.46     22788
weighted avg       0.72      0.85      0.78     22788

Confusion Matrix:
[[19358     0]
 [ 3430     0]]
AUC score:
0.5


c:\Technocolabs Internship\Prosper Project\Financial-Risk-Analysis-and-Prediction-of-online-P2P-Lending\venv\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Technocolabs Internship\Prosper Project\Financial-Risk-Analysis-and-Prediction-of-online-P2P-Lending\venv\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Technocolabs Internship\Prosper Project\Financial-Risk-Analysis-and-Prediction-of-online-P2P-Lending\venv\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: 

# ML Modeling using PCA dataset

In [34]:
dataset_pca = pd.read_csv("C:\Technocolabs Internship\Prosper Project\Financial-Risk-Analysis-and-Prediction-of-online-P2P-Lending\EDA_PCA_dataset.csv")

In [35]:
dataset_pca.head()

0         1         2         3         4         5         6  \
0 -4.706806 -1.093935  2.496602 -0.363543 -0.496769  3.829908  0.212925   
1 -4.882097 -0.933910  2.501757 -0.976342 -0.087196  3.619952 -0.608186   
2 -4.712364 -1.043991  1.518025 -1.035030  0.077261  3.504502 -0.468240   
3 -4.441806 -1.511071  1.469043 -0.427521  0.894490  3.424849 -0.964305   
4 -4.916132 -1.028381  2.477097 -0.971460 -0.192679  3.378130 -0.359368   

          7         8         9        10        11        12        13  \
0  0.218744  2.108790  0.891234 -0.884928 -0.333934  0.972919  0.212545   
1  1.097786  2.490542  0.835083 -1.087723 -0.159468  1.856734  0.395413   
2  1.177169  0.774237  0.930810 -1.197188 -0.065180  1.681108  0.391156   
3  1.119438  0.507444  1.250234 -1.344406  0.078951  2.826691  0.585382   
4  1.577417  2.517942  0.902052 -0.997856 -0.227580  1.622229  0.400885   

         14        15        16        17        18        19        20  \
0 -0.009062 -0.373314 -0.731456 -0.372348  0.159136  0.301718 -0.174700   
1 -0.091231 -0.656707 -1.088491 -0.198657  0.218885  0.164210  0.127286   
2 -0.141589 -0.610123 -1.098489 -0.009901  0.213829 -0.012411 -0.036102   
3 -0.228080 -0.796209 -1.424154  0.264955 -0.405423 -0.163212 -0.579545   
4 -0.068776 -0.458383 -0.998337 -0.156232  0.264540  0.182391  0.138062   

         21        22        23        24        25        26        27  \
0  1.893304 -1.278154 -1.518922 -1.104157  1.235884 -1.082835  0.482844   
1  2.453678 -1.157188 -1.658919 -0.919108  1.048985 -0.922913  0.501159   
2  2.164106 -1.115438 -1.067535 -1.409937 -1.057318 -1.383487  1.425040   
3  2.839717 -1.020801 -1.437837 -1.390610 -1.209624 -1.579333  1.148944   
4  2.277890 -1.177837 -1.657741 -0.935800  1.113018 -0.969815  0.681641   

         28        29        30        31        32        33        34  \
0 -0.082050  0.214366  0.174720 -1.444314 -0.365236 -1.085522  0.261368   
1 -0.219545  0.138607  0.139354 -1.132666  0.519238 -0.520896 -0.596439   
2 -0.309758  0.160085  0.170656 -1.786362  0.789292 -0.545390 -0.734297   
3 -0.091170  0.191427  0.321413 -1.918718 -0.037769 -0.955666 -0.317030   
4 -0.257115  0.120054  0.112544 -0.999868  0.433298 -0.582715 -0.427180   

         35        36        37        38        39        40        41  \
0  0.107997  0.109776 -0.794696 -0.180099  0.410118 -0.285390 -1.039791   
1 -0.283072 -0.117428 -1.096513  0.215753 -0.160075 -0.097847 -0.150250   
2 -0.223705 -0.559558  0.792998  0.246760  0.110836  0.093643  0.019628   
3  0.105029 -0.158512  1.108667  0.267252  0.559939 -0.214451 -0.719725   
4 -0.216912 -0.256146 -0.527990 -0.137086 -0.015230 -0.074576 -0.161426   

         42        43        44        45        46        47  IsDefault  
0 -0.752800 -0.243941  0.127581  0.198664 -0.275991  0.383254          0  
1 -0.267031 -0.157230 -0.027781  0.104809 -0.278525  0.324844          0  
2  0.573574  0.016317 -0.113583  0.081076 -0.249061  0.261768          0  
3  0.197562 -0.012854  0.222973 -0.125280 -0.247891  0.355171          0  
4 -0.262002 -0.117386 -0.041626  0.125339 -0.240471  0.283938          0

In [36]:
dataset_pca.shape

(113937, 49)

In [37]:
# Independent and dependent variables

X_pca = dataset_pca.drop(['IsDefault'], axis=1)
y_pca = dataset_pca['IsDefault']

In [38]:
# Splitting the dataset into train and test datasets

from sklearn.model_selection import train_test_split

X_train_pca, X_test_pca, y_train_pca, y_test_pca = train_test_split(X_pca, y_pca, test_size=0.2, random_state=42)

In [39]:
X_train_pca_upsampled, y_train_pca_upsampled = os.fit_resample(X_train_pca, y_train_pca)

print("The number of classes before oversampling: {}".format(Counter(y_train_pca)))
print("The number of classes after oversampling: {}".format(Counter(y_train_pca_upsampled)))

The number of classes before oversampling: Counter({0: 77569, 1: 13580})
The number of classes after oversampling: Counter({0: 77569, 1: 77428})


In [40]:
print("Resampled class distribution:", np.bincount(y_train_pca_upsampled))

Resampled class distribution: [77569 77428]


# Applying Logistic Regression

In [41]:
from sklearn.linear_model import LogisticRegression

log_reg_model_pca = LogisticRegression()

log_reg_model_pca.fit(X_train_pca_upsampled, y_train_pca_upsampled)

LogisticRegression()

In [42]:
y_pred_log_reg_pca = log_reg_model_pca.predict(X_test_pca)

In [43]:
# Evaluate the model
accuracy = accuracy_score(y_test_pca, y_pred_log_reg_pca)
print(f"Accuracy: {accuracy:.4f}")

# Print classification report
print("\nClassification Report:")
print(classification_report(y_test_pca, y_pred_log_reg_pca))

# Print confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test_pca, y_pred_log_reg_pca))

# Print confusion matrix
print("AUC score:")
print(roc_auc_score(y_test_pca, y_pred_log_reg_pca))

Accuracy: 0.4989

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.50      0.63     19358
           1       0.15      0.51      0.23      3430

    accuracy                           0.50     22788
   macro avg       0.50      0.50      0.43     22788
weighted avg       0.75      0.50      0.57     22788

Confusion Matrix:
[[9636 9722]
 [1696 1734]]
AUC score:
0.5016590273734396


# Naive Bayes

In [44]:
gnb_model_pca = GaussianNB()

In [45]:
gnb_model_pca.fit(X_train_pca_upsampled, y_train_pca_upsampled)

GaussianNB()

In [46]:
y_pred_gnb_pca = gnb_model_pca.predict(X_test_pca)

In [47]:
# Evaluate the model
accuracy = accuracy_score(y_test_pca, y_pred_gnb_pca)
print(f"Accuracy: {accuracy:.4f}")

# Print classification report
print("\nClassification Report:")
print(classification_report(y_test_pca, y_pred_gnb_pca))

# Print confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test_pca, y_pred_gnb_pca))

# Print confusion matrix
print("AUC score:")
print(roc_auc_score(y_test_pca, y_pred_gnb_pca))

Accuracy: 0.4279

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.40      0.54     19358
           1       0.15      0.61      0.24      3430

    accuracy                           0.43     22788
   macro avg       0.50      0.50      0.39     22788
weighted avg       0.75      0.43      0.50     22788

Confusion Matrix:
[[ 7659 11699]
 [ 1337  2093]]
AUC score:
0.5029272293688629


# LightGBM

In [48]:
lgb_model_pca = lgb.LGBMClassifier()

In [49]:
X_train_pca_upsampled.columns = X_train_pca_upsampled.columns.str.replace(r'[^\w]', '_', regex=True)

In [50]:
lgb_model_pca.fit(X_train_pca_upsampled, y_train_pca_upsampled)

[LightGBM] [Info] Number of positive: 77428, number of negative: 77569
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.025015 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 12240
[LightGBM] [Info] Number of data points in the train set: 154997, number of used features: 48
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.499545 -> initscore=-0.001819
[LightGBM] [Info] Start training from score -0.001819


LGBMClassifier()

In [51]:
# Prediction on test data
y_pred_lgbm_pca = lgb_model_pca.predict(X_test_pca)

In [52]:
# Evaluate the model
accuracy = accuracy_score(y_test_pca, y_pred_lgbm_pca)
print(f"Accuracy: {accuracy:.4f}")

# Print classification report
print("\nClassification Report:")
print(classification_report(y_test_pca, y_pred_lgbm_pca))

# Print confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test_pca, y_pred_lgbm_pca))

# Print confusion matrix
print("AUC score:")
print(roc_auc_score(y_test_pca, y_pred_lgbm_pca))

Accuracy: 0.7044

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.79      0.82     19358
           1       0.16      0.22      0.18      3430

    accuracy                           0.70     22788
   macro avg       0.50      0.50      0.50     22788
weighted avg       0.75      0.70      0.72     22788

Confusion Matrix:
[[15306  4052]
 [ 2685   745]]
AUC score:
0.5039410108205165


# Based on the above ML models training, the following results are achieved.

1. ML models training with Mutual Information (MI) dataset.

    a. Logistic Regression: Accuracy: 48%, AUC score: 0.5065

    b. Naive Bayes: Accuracy: 26%, AUC score: 0.4951
    
    c. LightGBM: Accuracy: 85%, AUC score: 0.5

2.  ML models training with With PCA dataset.

    a. Logistic Regression: Accuracy: 50%, AUC score: 0.5016

    b. Naive Bayes: Accuracy: 43%, AUC score: 0.5029

    c. LightGBM: Accuracy: 70%, AUC score: 0.5039

From the results, it’s evident that LightGBM demonstrates strong performance, especially in terms of accuracy. Comparing the datasets, the LightGBM model trained on features selected through Mutual Information stands out, achieving an accuracy of 85%. Therefore, the 'LightGBM model with Mutual Information features' is the best candidate for hyperparameter tuning to further optimize performance.

# Hyperparameter tuning.

In [53]:
from lightgbm import LGBMClassifier
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

In [54]:
# Initialize LightGBM classifier

lgb_model_final = LGBMClassifier(objective='binary', class_weight='balanced', boosting_type='gbdt')

In [55]:
# Set the Hyperparameters for LightGBM

param_grid = {
    'num_leaves': [31, 63, 127],
    'max_depth': [5, 7,10,20],
    'learning_rate': [0.01, 0.05],
    'n_estimators': [100, 150, 200, 500],
    'colsample_bytree': [0.8, 1.0],
    'subsample': [0.8, 1.0],
    'reg_alpha': [0, 0.1],
    'reg_lambda': [0, 0.1]
}

In [56]:
random_search = RandomizedSearchCV(estimator = lgb_model_final, 
                                   param_distributions = param_grid,
                                   n_iter=20, 
                                   scoring='roc_auc',   
                                   cv=5,
                                   random_state=42, 
                                   verbose=2, 
                                   n_jobs=-1
                                  )

random_search.fit(X_train_MI_upsampled, y_train_MI_upsampled)

Fitting 5 folds for each of 20 candidates, totalling 100 fits
[LightGBM] [Info] Number of positive: 77438, number of negative: 77569
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.018697 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 8877
[LightGBM] [Info] Number of data points in the train set: 155007, number of used features: 40
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=-0.000000
[LightGBM] [Info] Start training from score -0.000000


RandomizedSearchCV(cv=5,
                   estimator=LGBMClassifier(class_weight='balanced',
                                            objective='binary'),
                   n_iter=20, n_jobs=-1,
                   param_distributions={'colsample_bytree': [0.8, 1.0],
                                        'learning_rate': [0.01, 0.05],
                                        'max_depth': [5, 7, 10, 20],
                                        'n_estimators': [100, 150, 200, 500],
                                        'num_leaves': [31, 63, 127],
                                        'reg_alpha': [0, 0.1],
                                        'reg_lambda': [0, 0.1],
                                        'subsample': [0.8, 1.0]},
                   random_state=42, scoring='roc_auc', verbose=2)

In [57]:
# Best parameters and best model
best_params = random_search.best_params_
best_model = random_search.best_estimator_

print("Best parameters found: ", best_params)

Best parameters found:  {'subsample': 1.0, 'reg_lambda': 0.1, 'reg_alpha': 0, 'num_leaves': 127, 'n_estimators': 500, 'max_depth': 20, 'learning_rate': 0.05, 'colsample_bytree': 1.0}


In [58]:
# Train the model with the random search best parameters
best_model = lgb.LGBMClassifier(**random_search.best_params_, objective='binary', class_weight='balanced', boosting_type='gbdt')

best_model.fit(X_train_MI_upsampled, y_train_MI_upsampled)

[LightGBM] [Info] Number of positive: 77438, number of negative: 77569
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.017077 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 8877
[LightGBM] [Info] Number of data points in the train set: 155007, number of used features: 40
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=-0.000000
[LightGBM] [Info] Start training from score -0.000000


LGBMClassifier(class_weight='balanced', learning_rate=0.05, max_depth=20,
               n_estimators=500, num_leaves=127, objective='binary',
               reg_alpha=0, reg_lambda=0.1)

In [59]:
# Prediction with test data

y_predict_b = best_model.predict(X_test_MI)

In [60]:
# Evaluate the model
accuracy = accuracy_score(y_test_MI, y_predict_b)
print(f"Accuracy: {accuracy:.4f}")

# Print classification report
print("\nClassification Report:")
print(classification_report(y_test_MI, y_predict_b))

# Print confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test_MI, y_predict_b))

# Print confusion matrix
print("AUC score:")
print(roc_auc_score(y_test_MI, y_predict_b))

Accuracy: 0.8490

Classification Report:
              precision    recall  f1-score   support

           0       0.85      1.00      0.92     19358
           1       0.19      0.00      0.00      3430

    accuracy                           0.85     22788
   macro avg       0.52      0.50      0.46     22788
weighted avg       0.75      0.85      0.78     22788

Confusion Matrix:
[[19345    13]
 [ 3427     3]]
AUC score:
0.5001015392947432


# Saving the model

In [61]:
import pickle

In [62]:
filename = 'model.sav'

In [63]:
pickle.dump(best_model, open(filename, 'wb'))